src/controllers/PrefrenceController.ts

```diff
@@ -0,0 +1,229 @@
+          finalObject.cryptoCurrency = "-";
+          finalObject.fee = tx.meta.fee;
+          finalObject.type = inst.parsed.type;
+          finalObject.send = finalObject.from === selectedAddress;
+          finalObject.action = finalObject.send ? ACTIVITY_ACTION_SEND : ACTIVITY_ACTION_RECEIVE;
+          finalObject.decimal = decimals;
+          finalObject.totalAmountString = cryptoAmountToUiAmount(amount, decimals);
+          finalObject.logoURI = "";
+          finalObject.mintAddress = mint;
+        } else if (inst.program === "system") {
+          finalObject.from = inst.parsed.info.source;
+          finalObject.to = inst.parsed.info.destination;
+          finalObject.cryptoAmount = inst.parsed.info.lamports;
+          finalObject.cryptoCurrency = "SOL";
+          finalObject.fee = tx.meta.fee;
+          finalObject.type = inst.parsed.type;
+          finalObject.send = inst.parsed.info.source === selectedAddress;
+          finalObject.action = finalObject.send ? ACTIVITY_ACTION_SEND : ACTIVITY_ACTION_RECEIVE;
+          finalObject.decimal = 9;
+          finalObject.totalAmountString = lamportToSol(inst.parsed.info.lamports);
+          // finalObject.logoURI = default sol logo
+          // No converstion to current currency rate as the backend use transaction date currency rate
+        }
+      }
+    }
+    return finalObject;
+  });
+  return finalTxs;
+};
+
+export class ModPrefrenceController extends PreferencesController {
+  private getOverrideProviderConfig: NetworkController["getProviderConfig"];
+
+  private getOverrideConnection: NetworkController["getConnection"];
+
+  /**
+   * Creates a PreferencesController instance
+   *
+   * @param config - Initial options used to configure this controller
+   * @param state - Initial state to set on this controller
+   */
+  constructor(args: {
+    config?: SolanaPreferencesConfig;
+    state?: Partial<PreferencesState<ExtendedAddressPreferences>>;
+    signAuthMessage: KeyringController["signAuthMessage"];
+    getProviderConfig: NetworkController["getProviderConfig"];
+    getCurrentCurrency: CurrencyController["getCurrentCurrency"];
+    getConversionRate: CurrencyController["getConversionRate"];
+    getConnection: NetworkController["getConnection"];
+  }) {
+    super(args);
+    this.getOverrideProviderConfig = args.getProviderConfig;
+    this.getOverrideConnection = args.getConnection;
+  }
+
+  // Update with latest data from blockchain
+  async updateDisplayActivities(newActivities?: { [keyof: string]: SolanaTransactionActivity }): Promise<SolanaTransactionActivity[]> {
+    const address = this.state.selectedAddress;
+    const { chainId, blockExplorerUrl } = this.getOverrideProviderConfig();
+    const connection = this.getOverrideConnection();
+
+    // Get latest signature from blockchain for main Account
+    const signatureInfo = await connection.getSignaturesForAddress(new PublicKey(address), { limit: this.config.TX_LIMIT || 40 });
```

Comment:
`const [signatureInfo, parsedTransactions] = await Promise.all([
  connection.getSignaturesForAddress(...),
  connection.getParsedTransactions(...)
]);`

connection call can be parallelized

Reply:
we need the signature list before we can call parseTransactions